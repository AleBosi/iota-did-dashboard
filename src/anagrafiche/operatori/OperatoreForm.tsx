// src/anagrafiche/operatori/OperatoreForm.tsx

import React, { useState } from "react";
import { createOperatore } from "./OperatoreFactory";

type Props = {
  initial?: ReturnType<typeof createOperatore>;
  onSave: (o: ReturnType<typeof createOperatore>) => void;
  onCancel: () => void;
};

export default function OperatoreForm({ initial, onSave, onCancel }: Props) {
  const [form, setForm] = useState(
    initial || createOperatore()
  );

  function handleChange(e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) {
    const { name, value } = e.target;
    setForm(f => ({ ...f, [name]: value }));
  }

  function handleSubmit(e: React.FormEvent) {
    e.preventDefault();
    onSave(form);
  }

  return (
    <form onSubmit={handleSubmit} className="space-y-4 p-4">
      <input name="nome" value={form.nome} onChange={handleChange} placeholder="Nome" required className="input" />
      <input name="ruolo" value={form.ruolo} onChange={handleChange} placeholder="Ruolo" required className="input" />
      <input name="email" value={form.email} onChange={handleChange} placeholder="Email" required className="input" />
      <select name="stato" value={form.stato} onChange={handleChange} className="input">
        <option value="Attivo">Attivo</option>
        <option value="Sospeso">Sospeso</option>
        <option value="Cessato">Cessato</option>
      </select>
      <div className="space-x-3">
        <button type="submit" className="bg-green-600 text-white px-4 py-2 rounded">Salva</button>
        <button type="button" className="bg-gray-400 text-white px-4 py-2 rounded" onClick={onCancel}>Annulla</button>
      </div>
    </form>
  );
}
