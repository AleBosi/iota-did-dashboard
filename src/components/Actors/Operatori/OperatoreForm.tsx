import { useState } from "react";
import { Actor } from "../../../models/actor";

interface Props {
  onCreate: (operatore: Actor) => void;
}

// Funzione mock per generare un DID coerente (sostituisci con la tua se serve)
function generateDid(name: string) {
  return "did:iota:evm:operatore:" + name.trim().toLowerCase().replace(/\s+/g, "-") + "-" + Date.now();
}

export default function OperatoreForm({ onCreate }: Props) {
  const [name, setName] = useState("");

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (!name.trim()) return;
    const did = generateDid(name);
    const actor: Actor = {
      id: did,
      did,
      name: name.trim(),
      role: "operatore"
    };
    onCreate(actor);
    setName("");
  };

  return (
    <form onSubmit={handleSubmit} className="flex gap-2">
      <input
        value={name}
        onChange={e => setName(e.target.value)}
        placeholder="Nome operatore"
        className="border px-2 py-1 rounded"
      />
      <button
        type="submit"
        className="bg-blue-500 text-white px-4 py-1 rounded"
        disabled={!name.trim()}
      >
        Aggiungi
      </button>
    </form>
  );
}
