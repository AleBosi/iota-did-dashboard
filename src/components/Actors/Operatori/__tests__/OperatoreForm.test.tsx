import { render, screen, fireEvent } from "@testing-library/react";
import OperatoreForm from "../OperatoreForm";

describe("OperatoreForm", () => {
  const mockOnCreate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renderizza campo nome e bottone", () => {
    render(<OperatoreForm onCreate={mockOnCreate} />);
    expect(screen.getByPlaceholderText(/nome operatore/i)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /aggiungi/i })).toBeInTheDocument();
  });

  test("il bottone Ã¨ disabilitato se il nome Ã¨ vuoto o solo spazi", () => {
    render(<OperatoreForm onCreate={mockOnCreate} />);
    const button = screen.getByRole("button", { name: /aggiungi/i });
    expect(button).toBeDisabled();
    fireEvent.change(screen.getByPlaceholderText(/nome operatore/i), { target: { value: "  " } });
    expect(button).toBeDisabled();
  });

  test("il bottone si abilita se viene inserito un nome valido", () => {
    render(<OperatoreForm onCreate={mockOnCreate} />);
    const button = screen.getByRole("button", { name: /aggiungi/i });
    fireEvent.change(screen.getByPlaceholderText(/nome operatore/i), { target: { value: "Anna" } });
    expect(button).not.toBeDisabled();
  });

  test("chiama onCreate con il nome inserito e resetta il campo", () => {
    render(<OperatoreForm onCreate={mockOnCreate} />);
    const input = screen.getByPlaceholderText(/nome operatore/i);
    const button = screen.getByRole("button", { name: /aggiungi/i });

    fireEvent.change(input, { target: { value: "Anna" } });
    fireEvent.click(button);

    expect(mockOnCreate).toHaveBeenCalledTimes(1);
    expect(mockOnCreate).toHaveBeenCalledWith({ name: "Anna" });
    expect(input).toHaveValue("");
  });

  test("non chiama onCreate se si prova a inviare con campo vuoto", () => {
    render(<OperatoreForm onCreate={mockOnCreate} />);
    fireEvent.click(screen.getByRole("button", { name: /aggiungi/i }));
    expect(mockOnCreate).not.toHaveBeenCalled();
  });

  // Snapshot base
  test("snapshot", () => {
    const { asFragment } = render(<OperatoreForm onCreate={mockOnCreate} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
