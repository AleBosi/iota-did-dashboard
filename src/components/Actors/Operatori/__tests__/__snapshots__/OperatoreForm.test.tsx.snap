// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`OperatoreForm snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex gap-2"
  >
    <input
      class="border px-2 py-1 rounded"
      placeholder="Nome operatore"
      value=""
    />
    <button
      class="bg-blue-500 text-white px-4 py-1 rounded"
      disabled=""
      type="submit"
    >
      Aggiungi
    </button>
  </form>
</DocumentFragment>
`;
